File: 044 VB Attributes/MyApp.vb
' 44 VB Attributes
' Essai d'utilisation des attributs en VB
' 2001-02-18    PV
' 2001-08-18    PV Beta2
' 2006-10-01    PV  VS2005
' 2012-02-25	PV  VS2010

<AttributeUsage(AttributeTargets.Class)> Public Class MonAttribut
    Inherits Attribute
    Private ReadOnly iPriv As Integer
    Private sInfo As String

    Public Sub New(iVal As Integer)
        iPriv = iVal
        sInfo = ""
    End Sub

    Public Property Info() As String
        Get
            Return sInfo
        End Get
        Set(Value As String)
            sInfo = Value
        End Set
    End Property

    Public ReadOnly Property IFlags() As Integer
        Get
            Return iPriv
        End Get
    End Property

End Class

<MonAttribut(1, Info:="Info de MaClasse1")> Class MaClasse1

End Class

<MonAttribut(7)> Class MaClasse2

End Class

Class MyApp

    Public Shared Sub Main()
        Dim o1 As New MaClasse1()
        Dim o2 As New MaClasse2()

        Zap(o1)
        Zap(o2)

        Console.ReadLine()
    End Sub

    Private Shared Sub Zap(o As Object)
        Dim tob As Type     ' Type de l'objet
        tob = o.GetType()
        Console.WriteLine(tob.Name)

        Dim tat As Type     ' Type de l'attribut
        tat = GetType(MonAttribut)

        Dim m As MonAttribut
        m = CType(tob.GetCustomAttributes(tat, False)(0), MonAttribut)
        Console.WriteLine("{0}, {1}", m.IFlags, m.Info)
    End Sub

End Class

File: FW4.8/044 CS Attributes/MyApp.cs
// 44 C# Attributes
// Essai d'utilisation des attributs en C#
// 2001-02-18   PV
// 2001-08-18   PV Beta2
// 2006-10-01   PV  VS2005
// 2012-02-25   PV  VS2010

using System;

[AttributeUsage(AttributeTargets.Class)]
public class MonAttribut : System.Attribute
{
    private readonly int iPriv;
    private string sInfo;

    public MonAttribut(int iVal)
    {
        iPriv = iVal;
        sInfo = "";
    }

    public string Info
    {
        get => sInfo;
        set => sInfo = value;
    }

    public int IFlags => iPriv;
}

[MonAttribut(1, Info = "Info de MaClasse1")]
internal class MaClasse1
{
}

[MonAttribut(7)]
internal class MaClasse2
{
}

internal class MyApp
{
    public static void Main()
    {
        var o1 = new MaClasse1();
        var o2 = new MaClasse2();

        Zap(o1);
        Zap(o2);

        Console.ReadLine();
    }

    private static void Zap(object o)
    {
        Type t = o.GetType();
        Console.WriteLine(t.Name);

        var m = (MonAttribut)t.GetCustomAttributes(typeof(MonAttribut), false)[0];
        Console.WriteLine("{0}, {1}", m.IFlags, m.Info);
    }
}